index.js

```diff
@@ -17,19 +18,30 @@ function getFilenameFromMime(name, mime) {
 	return `${name}.${exts[0].ext}`;
 }
 
-function registerListener(session, options, cb = () => {}) {
-	const downloadItems = new Set();
-	let receivedBytes = 0;
-	let completedBytes = 0;
-	let totalBytes = 0;
-	const activeDownloadItems = () => downloadItems.size;
-	const progressDownloadItems = () => receivedBytes / totalBytes;
-
-	options = Object.assign({
-		showBadge: true
-	}, options);
+const sessionListenerMap = new Map();
+const handlerMap = new Map();
+const downloadItems = new Set();
+let receivedBytes = 0;
+let completedBytes = 0;
+let totalBytes = 0;
+const activeDownloadItems = () => downloadItems.size;
+const progressDownloadItems = function (item) {
+	if (item) {
+		return item.getReceivedBytes() / item.getTotalBytes();
+	}
+	return receivedBytes / totalBytes;
+};
 
+function registerListener(session) {
 	const listener = (e, item, webContents) => {
+		const url = decodeURIComponent(_.first(item.getURLChain()));
```

Comment:
The URL created is different than handlerMap.set(decodeURIComponent(url) ?

We should have a utility function to create the consistent key

Reply:
we can not use same function because handleMap.set use download url - first time
When start download, the downloader will check the url & follow to new url if has redirect response.
We just have 1 way to get original url is get from item.getUrlChain() - first item

Reply:
Not quite sure what you meant, the way hashmap works is that you use the same key, if we cannot use the same key, then there's no point using a hashmap?

Reply:
value of item.getUrlChain() (first item) is equal with url that set in L144

Reply:
That's not very clear that it'll be equal? Create a function and use it consistently?

Reply:
I think it's difference context. So we can not use same way to get it. The first when we set it, We know exactly it's a key. the second time when we use it. The context of code is in download handler. We just have only way to get it is based on `item.getChainUrl`

Reply:
If that's the case, let's break it up into two variables, one for key and one for url to avoid confusion?